### DISCUSSION TITLE
Unneeded suggestions.

### DISCUSSION THREAD

**Discussion by PedroMirM** [littleK0i/SnowDDL#57] [Q&A]
I'm looking at using Snowddl at my company, but I'm confused. Sometimes, Snowddl suggests changes to its "plan" that don't seem needed. 

Like, I have two tables: Table A and Table B. If I change Table B, it tells me to update a comment in Table A, even if I didn't change Table A. This also happens with views. Even if I don't touch them or their tables, it tells me to recreate the views. Can someone explain why? 🤔 

Thanks.

**Comment by littleK0i** [comment]
Could you attach YAML configs and output from `SHOW TABLES` / `SHOW VIEWS` by any chance? I'll take a look.

About tables: it is possible that Snowflake started to "normalize" or modify comments while saving. SnowDDL compares exact comment from `SHOW TABLES` with comment in YAML. It can only product `SET COMMENT` on table if these string do not match: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/resolver/table.py#L214-L221

About views: existing views are re-created only if view text was changed, or view is invalid and cannot be queried. Here is the entire logic, which is quite short: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/resolver/view.py#L50-L85

Is it possible that some other script or person tries to update objects in the background?

**Reply by PedroMirM** [reply]
Just like you suggested, `Terraform` was changing some views. Thanks for the hint! :))

**Reply by PedroMirM** [reply]
Well, no, in some cases Terraform isn't doing anything and I still have the same problem :S
[show views.csv](https://github.com/littleK0i/SnowDDL/files/12672226/show.views.csv)
[show tables.csv](https://github.com/littleK0i/SnowDDL/files/12672228/show.tables.csv)

**snowddl/single_db_config/SNOWFLAKE_TEST_DOMAIN/snowflake/test/shared_staging/view/BASE_SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST.sql**
`select raw from TEST.SHARED_SOURCE.SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST_VERSION_LIVE_1;`

**snowddl/single_db_config/SNOWFLAKE_TEST_DOMAIN/snowflake/test/shared_staging/view/BASE_SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST.yaml**
```
columns:
  RAW: "raw"

text: !include BASE_SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST.sql

comment: base view for SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST
```

And for some reason, Snowddl is trying to recreate these veiws everty time...
```
2023-09-20 10:49:17.216 - INFO - ---
2023-09-20 10:49:20.921 - INFO - Resolved VIEW [TEST.SHARED_ANALYTICS.EMPLOYEES_DEPARTMENTS]: REPLACE
2023-09-20 10:49:20.921 - INFO - Resolved VIEW [TEST.SHARED_STAGING.BASE_SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST]: REPLACE
2023-09-20 10:49:21.285 - INFO - Executed 0 DDL queries, Suggested 4 DDL queries
--- Suggested DDL ---



CREATE OR REPLACE VIEW "TEST"."SHARED_STAGING"."BASE_SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST"
(
      "RAW" COMMENT 'raw'
)
COPY GRANTS
AS
select raw from TEST.SHARED_SOURCE.SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST_VERSION_LIVE_1;
;

COMMENT ON VIEW "TEST"."SHARED_STAGING"."BASE_SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST" IS 'base view for SRC_AWS_S3_DATAPRODUCTION_ANONYMIZEDDATASTOREDEVELOPMENT_SEARCH_SYSTEMS_DEV_ANSWER_SERVICE_US_REQUEST';




```

**Reply by littleK0i** [reply]
Could you try to remove trailing `;` from view SQL text? It is the only suspicious part. Everything else is trivial.

I've checked response from `SHOW VIEWS` and SQL text suggested by SnowDDL, and it seems to be exactly the same.

**Reply by mirkanos** [reply]
I've removed the `;` you mentioned, but it still recreates all the views with their comments. It's strange because I haven't changed anything about them as I said...
As you pointed out, the code is straightforward, https://github.com/littleK0i/SnowDDL/blob/master/snowddl/resolver/view.py#L50-L85

Is there a way to debug that part of the code to understand the condition that makes Snowddl decide to recreate the view?
I can't think of anything else. I'm concerned that while it might not be an issue now since we don't have many views, it could become a problem in the future if these views are continually recreated, especially as we have more of them...

**Reply by littleK0i** [reply]
Sure, you can install snowddl in "developer" mode.

1. Download repo files as ZIP: https://github.com/littleK0i/SnowDDL/archive/refs/heads/master.zip
2. Extract into any directory;
3. Open this directory in terminal (e.g. `cd path/to/snowddl`);
4. Run `pip install -e .`

Now `snowddl` package points to your local code directory instead of WHL downloaded from PyPI.

You may add debug lines to this method and see how it goes: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/resolver/view.py#L50-L85

Basic `print` works fine. You may also use default SnowDDL logger from `self.engine.logger`.

E.g.
```python
...
self.engine.logger.info("My debug")
...
```

**Reply by littleK0i** [reply]
The problem might be happening because of this part:

```python
            try:
                # ... and it is possible to query view (underlying objects were not changed)
                self.engine.describe_meta("SELECT * FROM {full_name:i}", {
                    "full_name": bp.full_name,
                })
            except SnowDDLExecuteError as e:
                # TODO: add checks for specific errors?
                pass
```

This code is used to check validity of views. It run method [`describe`](https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#describe) method from Python connector. If this method returns list of columns, view is valid. However, if it raises an exception, it means view might be invalid and should be re-created.

I suspect you might be getting a tricky exception, which is not related to view or table structure.

I suggest to start debugging by dumping details of potential exception happening here.

Thank you!
